Webpack5/Vite: Fix sourcemaps

valentinpalkovic opened the pull request:
Closes https://github.com/storybookjs/storybook/issues/26954
Closes https://github.com/storybookjs/storybook/issues/26653
Closes https://github.com/ni/nimble/issues/2066

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/contribute

-->

## What I did

Fixes sourcemaps for the Webpack5 builder.

Several loaders in the Webpack5 loader chain were not processing source maps properly, resulting in either incorrect or incomplete mapping. 

Second, I have configured unplugin for each builder separately so that sourcemaps can be handled individually.

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

1. Create a Nextjs sandbox
2. Open Storybook in your browser
3. Throw an error in the onclick handler within the button story. Click the button and see whether it maps to the correct line (just line mapping works because the default sourcemap option in Webpack5 is set to `cheap-module-source-map` (A SourceMap without column-mappings that simplifies loader Source Maps to a single mapping per line)
4. Run Next.js in babel mode by creating a `.babelrc` file with the following content:

```
{
  "presets": ["next/babel"],
  "plugins": []
}
```
5. Repeat step 3

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->
This pull request has been released as version `0.0.0-pr-27171-sha-a9f1e1db`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-27171-sha-a9f1e1db sandbox` or in an existing project with `npx storybook@0.0.0-pr-27171-sha-a9f1e1db upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-27171-sha-a9f1e1db`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-27171-sha-a9f1e1db) |
| **Triggered by** | @valentinpalkovic |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`valentin/fix-webpack5-sourcemaps`](https://github.com/storybookjs/storybook/tree/valentin/fix-webpack5-sourcemaps) |
| **Commit** | [`a9f1e1db`](https://github.com/storybookjs/storybook/commit/a9f1e1db7f996f63290744d1015fd660141c13ed) |
| **Datetime** | Fri May 24 10:37:30 UTC 2024 (`1716547050`) |
| **Workflow run** | [9222633276](https://github.com/storybookjs/storybook/actions/runs/9222633276) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=27171`_
</details>
<!-- CANARY_RELEASE_SECTION -->

valentinpalkovic committed:
Webpack5: Fix sourcemaps

valentinpalkovic committed:
Deprecate @storybook/csf-plugin and remove references

valentinpalkovic committed:
Fix addon-docs

valentinpalkovic committed:
Fix addon-docs

valentinpalkovic committed:
Vite: Fix sourcemapping for CSF files

valentinpalkovic committed:
Small fixes

valentinpalkovic committed:
Merge branch 'next' into valentin/fix-webpack5-sourcemaps

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into valentin/fix-webpack5-sourcemaps

yannbf commented on code/addons/docs/package.json:
```diff
@@ -36,11 +36,7 @@
-      "require": "./dist/preset.js",
-      "import": "./dist/preset.js"
-    },
+    "./preset": "./dist/preset.js",
```
are these changes (and on line 50) necessary? why?

valentinpalkovic replied:
Yes, they are. `addon-docs` wasn't using the `addon-bundle` script, but the old build script instead. I couldn't build addon-docs with the old script, though, because I was using an import to `fs` and the build script assumed that I would build for the browser, which isn't the case for a Webpack5 loader. That's the reason why I have upgraded addon-docs to use the addon-bundle script instead.

yannbf commented on code/builders/builder-vite/src/plugins/csf-plugin.ts:
```diff
@@ -10,6 +14,34 @@ export async function csfPlugin(config: Options): Promise<Plugin> {
+    name: 'vite-plugin-csf',
+    enforce: 'pre',
+    async transform(code, id) {
+      if (!STORIES_REGEX.test(id)) {
```
This might be silly but if anyone uses any other pattern they want for story file names, this will bail. Not that they should do that, as most of Storybook's tools only detect stories/story format anyway. Just an FYI!

valentinpalkovic replied:
That's correct. I have just copied over this code from the csf-plugin. We should revisit this at some point and read the users story glob instead and transform it to a regex pattern.

yannbf commented on code/frameworks/nextjs/src/swc/loader.ts:
```diff
@@ -31,7 +31,6 @@ export const configureSWCLoader = async (
     test: /\.((c|m)?(j|t)sx?)$/,
     include: [getProjectRoot()],
     exclude: [/(node_modules)/, ...Object.keys(virtualModules)],
-    enforce: 'post',
```
now that the order of the loader is changed, are there potential side effects to this?

valentinpalkovic replied:
I have adjusted the loader so that it behaves like the Babel loader. There might be side effects, but I am aware of none right now.

yannbf reviewed (approved):
LGTM, but I'm not sure about potential downsides/side effects. I'd say we should properly test with a canary as we don't have any tests in place regarding sourcemaps (I don't even know how we'd be testing this in an automated way)

valentinpalkovic commented on code/lib/csf-plugin/README.md:
```diff
@@ -1,4 +1,4 @@
-# CSF Plugin
+# [DEPRECATED] CSF Plugin
 
 The CSF plugin reads CSF files and enriches their content via static analysis.
```
So I did a GitHub search, and indeed, it is used in a couple of projects. That said, I think it makes sense not to deprecate it but rather to move the custom Webpack loader and Vite Plugin back into `@storybook/csf-plugin` in Storybook 9 with some breaking change (we don't provide a Webpack plugin, but a Webpack loader instead).

valentinpalkovic committed:
Merge branch 'next' into valentin/fix-webpack5-sourcemaps

shilman commented on scripts/tasks/sandbox-parts.ts:
```diff
@@ -219,10 +219,18 @@ function addEsbuildLoaderToStories(mainConfig: ConfigFile) {
-        })),
+        ...config.module.rules.map(rule => {
+
+          // The csf loader is responsible to add metadata to CSF files (e.g. extracting JSON comments)
```
Do we need to do something similar for Vite?

valentinpalkovic replied:
I do not even know why we are excluding all rules per default for the sandboxes and only for Webpack. I have no context for this. We don't have to do something similar for Vite.

shilman reviewed (commented):
Is there some way to do a single automated test of the toolchain in both Webpack and Vite, e.g.

```js
import { Button } from './Button';
/** use jsdoc to to exercise the csf enhancement */
export default { component: Button };
/** again */
export const Basic = {}
```

So we can (1) verify that it's working properly, and more importantly (2) verify that it doesn't regress later

valentinpalkovic committed:
Merge branch 'next' into valentin/fix-webpack5-sourcemaps

valentinpalkovic committed:
Merge branch 'next' into valentin/fix-webpack5-sourcemaps

valentinpalkovic committed:
Revert "Deprecate @storybook/csf-plugin and remove references"

This reverts commit 19fb74d6cb9e016e6067352f83b590281f0f5efe.

valentinpalkovic commented:
@shilman I had the idea to use a source map visualization like this one: https://github.com/evanw/source-map-visualization?tab=readme-ov-file, but instead it analyzes a story and shows up the correct source map mapping.

I tried it out for an hour but I would need definitely more time to get it somehow working. If it would work, we can just easily snapshot things :)

valentinpalkovic committed:
Rework csf-plugin to guarantee sourcemap functionality in all builders

valentinpalkovic committed:
Revert: Small fixes

valentinpalkovic committed:
Revert "Vite: Fix sourcemapping for CSF files"

This reverts commit 04378b75a1d6f19df2bd16e55ab8ea04343116f5.

valentinpalkovic committed:
Revert "Fix addon-docs"

This reverts commit 0200052e0976a08555d32024831b4750915d6e60.

valentinpalkovic committed:
Revert "Fix addon-docs"

This reverts commit 61ecef8a3990657e4c4cdea56566cec776c96811.

valentinpalkovic committed:
Partially Revert "Webpack5: Fix sourcemaps"

This reverts commit 9667b1e6c9ddadbe9ad1b16b5c41f57f21aad49e.

valentinpalkovic committed:
Fix sourcemaps across loaders and builders

valentinpalkovic committed:
Merge branch 'next' into valentin/fix-webpack5-sourcemaps

valentinpalkovic committed:
Merge branch 'next' into valentin/fix-webpack5-sourcemaps

valentinpalkovic committed:
Fix csf plugin for webpack and rspack

valentinpalkovic committed:
Remove includeContent option in magic string source map generation

Vite/Rollup only needs the "mappings" field for sourecemaps and ignores sourcesContentarrays. Therefore, there is no need to generate "sourcesContent" by setting the "includeContent" flag to true for magicString sourcemaps generation

valentinpalkovic committed:
Fix sourcemaps for Vite

valentinpalkovic committed:
Fix sourcemaps for Webpack5

sag1v commented:
@shilman Is this fix available in `v7.6.20`? I'm experiencing a similar issue of not seeing some of the source maps (webpak5)

valentinpalkovic commented:
Hi @sag1v 

this isn’t patches back to Storybook 7. Please upgrade to 8 and let us know if you have difficulties with the upgrade.

sag1v commented:
Upgrading to v8 is a bit of a blocker for us atm as we still use yarn v1 in this specific project unfortunately.

Is there something i can do in "user land" to fix it?

valentinpalkovic commented:
Although we don’t explicitly support yarn 1, it should still work with Storybook 8. Can you elaborate your issues in a new GitHub issue? I will try to help you upgrading to SB 8.

I don’t think you can fix this issue from your side unfortunately.

sag1v commented:
I will try to make time for it later on this week. Thanks 🙏🏽
